src/replication.c

```diff
@@ -2873,6 +2873,9 @@ void replicationUnsetMaster(void) {
     disconnectSlaves();
     server.repl_state = REPL_STATE_NONE;
 
+    /* Release slaveKeysWithExpire dict, since master doesn't need it. */
+    flushSlaveKeysWithExpireList();
```

Comment:
IIRC when i proposed a similar fix, Salvatore said that it would be wrong, since we still need to expire (delete) the temporary keys that were created while the instance was a replica (rather than persist them, which is what this change would do.

see https://github.com/redis/redis/pull/6219#issuecomment-510843953

Reply:
>about flushing the table when the replica is promoted to master: one problem that may occur is that the temporary keys are there and yet not expired. Later the new master is demoted to replica again after a failover, doing just a partial resynchronization: in that case, we'll leak the keys. So maybe it makes more sense to just do that when the DB is emptied completely?

that won't happen, because when a replica is promoted to master, we changed the repl-id, and then the master cannot perform a partial sync with its replica when failover.

Reply:
Oh, Salvatore means the replica with expire keys already has replica, that's a problem, thinking...

Reply:
this is all a big pile of mess and bugs, but let's put aside for a moment what happens when someone tries to sync from the promoted replica.

if we merge the flush we both proposed, it means that these keys that were volatile while the instance was a replica, they'll become persisted when it is promoted, right?

if some new replica then syncs from the promoted replica and stays up for long enough, eventually all these keys will get expired, and DELs will be sent to the replica.

disclaimer: i didn't look into this mechanism again when typing these lines to confirm it does what i think it does.. i rather invest my time elsewhere and not suffer the nightmares it'll induce.

writable replicas are a mess. but writable replicas with volatile keys is a mess on a different scale :smile:

Reply:
>if we merge the flush we both proposed, it means that these keys that were volatile while the instance was a replica, they'll become persisted when it is promoted, right?

I think the scenario is a replica with special expire keys promote to master and then (before the expire keys deleted) demote to replica again (partial sync succeed). then these expires keys become persisted.

not only the expire keys have the problem but also all write commands nor from master would cause the problem, because of the partial sync, for example:

A->B->C, B is a writable-replica have inconsistency date with master A, after B promoted to master, C is still inconsistency with master B, only full-resync can fix it.

>i rather invest my time elsewhere and not suffer the nightmares it'll induce.
>writable replicas are a mess. but writable replicas with volatile keys is a mess on a different scale 😄

I totally agree, I don't even want waste time to fix it.